Molecular Mechanics Induction Energy Calculator

shehan807 opened the pull request:
## Description
**The goal of this program is to simplify the calculation of the induction energy component using a SAPT-based, ab initio polarizable molecular dynamics force field function form.** The theoretical formalism is elaborated in [Shehan's U_pol Repo](https://github.com/shehan807/U_pol). In essence, each relevant atom, considered a "core" atom, is paired with a "shell" or Drude particle through which induced-dipole/induced-dipole interactions manifest. The added functions have been verified to reproduce polarization energies in [OpenMM](https://github.com/openmm/openmm) with the advantage of a Drude SCF optimization scheme implemented using [JAX](https://github.com/jax-ml/jax). 

## Todos
Notable points that this PR has either accomplished or will accomplish.
  - [ ] match input/output schemes between nmer dictionary and Uind function in force_field.py
  - [ ] verify large scale dimer/trimer energies for imidazole, pyrazole, and acetic acid  

## Questions
- [ ] Where would the drudeOpt function go? 
- [ ] What is the best way to atom type Psi4 atoms and/or create Drude particle properties based on conventional SAPT-FF atom typing schemes?

## Status
- [ ] Ready to go

loriab commented on crystalatte/plugins/force_fields.py:
```diff
@@ -1,6 +1,468 @@
+from jaxopt import NonlinearCG
+import jax.numpy as jnp
+import jax
+from jax import jit
```
There's a lot of extra dependencies with this, so I think it should go into its own file

shehan807 replied:
Done — moved all the JAX/Drude polarization code into its own module, `crystalatte/plugins/polarization_ff.py

loriab commented on crystalatte/plugins/__init__.py:
```diff
@@ -1,4 +1,5 @@
 try:
+    from . import openmm_utils
```
You'll want these as separate blocks. try/import-A/except/pass  try/import-B/except/pass so the attempts are independent. What's openmm_utils — I don't see it here?

shehan807 replied:
openmm_utils was removed, and now this line is deleted.

loriab commented on crystalatte/plugins/force_fields.py:
```diff
@@ -1,6 +1,468 @@
+    - 1/4πε₀ is the Coulomb constant
+    - Qi_shell, Qj_shell are the Drude charges on shared molecules 
+    - R_ij is the distance matrix between sites i and j
+    - D_i, D_j are the displacements of Drude particles from parent atom nuclei
```
thanks for the nice docstrings

loriab commented on examples/polarization_ff/imidazole/imidazole_outputs/s4machinefile-79897:
```diff
@@ -0,0 +1,6 @@
+ds1
```
there's a lot of files added in this PR that maybe just got caught by a generic `git add`. I'd go through and remove any of the molecule or output files that aren't going to be used in a test or have a definite purpose. s4machinefiles can definitely go :-)

shehan807 commented:
Hi @loriab, thanks for your patience on this! I've addressed all your comments, trimmed the PR down to just the essential functions under `crystalatte/plugins/`, and isolated the heavy JAX/Drude deps into their own module (`polarization_ff.py`) behind guarded imports so they don't load with the base package. Added a regression test and a runnable example too.

CC'ing @Awallace3 — you might want to review `polarization_energy_function` wiring, which lets stock crystalatte dispatch to a custom energy function like the polarization FF.
